### ISSUE TITLE
Loading packages from custom URLs

### ISSUE THREAD

**Issue by rth** [iodide-project/pyodide#121]
Currently, included packages and their dependencies are defined in a global `packages.json`. Independently of how this mechanism evolves in the future (e.g. advanced versioning, etc as outlined in the [wiki](https://github.com/iodide-project/pyodide/wiki/Packaging)), additionally, it could be useful to support import of custom packages from user defined URLs,
```py
# from JS
pyodide.loadPackage('https://some_url/new_package.js')

# from Python
import new_package
```
A typical use case is when user would build some custom package locally, and would like to load it in pyodide. This is a feature that was suggested by [**klaus**](https://lab.nexedi.com/u/klaus) after having used pyodide internally at [Nexedi](https://github.com/Nexedi).

Opening this as a separate issue, as this discussion is a bit orthogonal to https://github.com/iodide-project/pyodide/issues/38 and how packages are actually built. The possibly relevant part of https://github.com/iodide-project/pyodide/issues/38 is, @mdboom 's comment,

>  The tricky bit is that the packages.json is currently monolithic, and we probably don't want to require these custom packages to modify it. The advantage of that approach, is that loadPackage can figure out all of the dependencies up front and load them in parallel -- i.e., it doesn't have to load a package first before determining its dependencies and loading those, which is unfortunately the pip/setuptools model. However, we could add dependency metadata to the pyodide package itself in its .js file, such that you could determine the dependencies of a custom package once loading it.

I think, this could be achieved with the following steps,
 - [x] update `pyodide.loadPackage` to optionally allow loading packages from custom urls (without handling dependencies)
 - [ ] optionally include dependencies in .js files and
 - [x] make `.js` and `.data` files relocatable. Currently `.js` contains the full path to `.data` where it was built, e.g.
   ```js
   var PACKAGE_NAME="/home/mdboom/Work/builds/compiling/pyodide/packages/numpy/build/numpy.data";
   ```
   at the moment, I'm not sure if it's an issue, but finding a way around that might still be preferable.

WDYT @mdboom ?

**Comment by mdboom** [comment]
I think this makes sense as a good first step toward independent compilation of packages.  I think adding a list of dependencies to the `.js` file is probably fine, despite needing more round trips.  We can build something that works and worry about performance later, if we have to.

Possibly, for simplicity's sake, we'll want to do this with the "built-in" packages and ditch the `packages.json` file.

There's the question of how the dependencies are specified.  I think the general idea should be that it calls `loadPackage` recursively on each of the strings in an array, such that if you said `matplotlib`, it would load it from the "core" pyodide packages set as it does now, but you could also specify `http://mydomain.com/my-special-matplotlib.js`.  Maybe some day we'll want shortcuts to certain repos like `pyodide-forge:matplotlib`, but that should be doable on top of this.

Finally, what are the unique keys to determine if a package is already loaded?  It should probably be the resulting python package name (i.e. `matplotlib`), not the full URL to where the package came from, since it's impossible to load two python packages called `matplotlib` at the same time, anyway.  I don't think the `.js` files contain that information already, so we'll need to add that as well.  To avoid confusion, we probably want to present a warning if the user tries to load two packages with the same Python package name from two different URLs (but silently do nothing if loading the same package from the same URL).

I wonder whether we want the `.js` in the URL -- the `.js`/`.data` pair is kind of an implementation detail.  Maybe we need special extensions for Pyodide packages?  (`.pypkg`, `pypkgdata` ??)  That can probably be addressed later, though.

I didn't realize the full path was making it into the `.js` files.  But it doesn't look like that variable is used for anything other than as a unique identifier to see if the file system overlay is already loaded.  So while I agree it's not ideal, it doesn't really prevent the packages from being relocatable (we post them as-is to github pages and they work just fine).  I've made #122 to track that independently.

**Comment by rth** [comment]
Thanks for the detailed feedback and all the different ideas! I'll work on proposing a few incremental changes to make this possible.

**Comment by rth** [comment]
In a sense packages were always requested as a separate file, but they needed to be located on the same base URL as the Python base. This allows to load them from any URL.

**Comment by rth** [comment]
Closing as mostly resolved. For pure python packages https://github.com/iodide-project/pyodide/issues/642 might be a more reliable way forward.
